javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/IndyModuleRegistry.java

```diff
@@ -0,0 +1,121 @@
+
+  public static InstrumentationModuleClassLoader getInstrumentationClassloader(
+      String moduleClassName, ClassLoader instrumentedClassloader) {
+    InstrumentationModule instrumentationModule = modulesByName.get(moduleClassName);
+    if (instrumentationModule == null) {
+      throw new IllegalArgumentException(
+          "No module with the class name " + modulesByName + " has been registered!");
+    }
+    return getInstrumentationClassloader(instrumentationModule, instrumentedClassloader);
+  }
+
+  private static synchronized InstrumentationModuleClassLoader getInstrumentationClassloader(
+      InstrumentationModule module, ClassLoader instrumentedClassloader) {
+
+    Cache<ClassLoader, WeakReference<InstrumentationModuleClassLoader>> cacheForModule =
+        instrumentationClassloaders.computeIfAbsent(module, (k) -> Cache.weak());
+
+    WeakReference<InstrumentationModuleClassLoader> cached =
+        cacheForModule.get(instrumentedClassloader);
+    if (cached != null) {
+      InstrumentationModuleClassLoader cachedCl = cached.get();
+      if (cachedCl != null) {
+        return cachedCl;
+      }
+    }
+    // We can't directly use "compute-if-absent" here because then for a short time only the
+    // WeakReference will point to the InstrumentationModuleCL
+    InstrumentationModuleClassLoader created =
+        createInstrumentationModuleClassloader(module, instrumentedClassloader);
+    cacheForModule.put(instrumentedClassloader, new WeakReference<>(created));
+    return created;
+  }
+
+  private static InstrumentationModuleClassLoader createInstrumentationModuleClassloader(
+      InstrumentationModule module, ClassLoader instrumentedClassloader) {
+
+    Set<String> toInject = new HashSet<>(InstrumentationModuleMuzzle.getHelperClassNames(module));
+    // TODO (Jonas): Make muzzle include advice classes as helper classes
+    // so that we don't have to include them here
+    toInject.addAll(getModuleAdviceNames(module));
+
+    ClassLoader agentOrExtensionCl = module.getClass().getClassLoader();
+    Map<String, ClassCopySource> injectedClasses =
+        toInject.stream()
+            .collect(
+                Collectors.toMap(
+                    name -> name, name -> ClassCopySource.create(name, agentOrExtensionCl)));
+
+    return new InstrumentationModuleClassLoader(
+        instrumentedClassloader, agentOrExtensionCl, injectedClasses);
+  }
+
+  public static void registerIndyModule(InstrumentationModule module) {
+    if (!module.isIndyModule()) {
+      throw new IllegalArgumentException("Provided module is not an indy module!");
+    }
+    String moduleName = module.getClass().getName();
+    if (modulesByName.putIfAbsent(moduleName, module) != null) {
+      throw new IllegalArgumentException(
+          "A module with the class name " + moduleName + " has already been registered!");
+    }
+  }
+
+  private static Set<String> getModuleAdviceNames(InstrumentationModule module) {
```

Comment:
Instead of doing this, muzzle could be improved to also store the advice classes' names during compile time.
And, when we're ready to retire the inlining instrumentation modules, include advice classes in `getMuzzleHelperClassNames()`

Reply:
I've added this todo for this:
https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/9382/files/4fe49e3f64e6dc20e05d057add1b18f84f280fb4#diff-09f4593d6f1ec7855d7931c28ac7a934104886543cdde7531a03cc6dd02a3017R77

I wanted to keep this out of this change because I didn't another component (muzzle) in this already big PR.
Are you okay with just the `todo` or should I adapt muzzle in this PR?

Reply:
I think a TODO is fine for now, I'd definitely want it done separately.